### PR TITLE
Refactor Take 1

### PR DESCRIPTION
**Pull request by VasanthaKasirajan3008**

### CODE REVIEW

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        private const string Cookies = "Cookies";        
+        private const string Staff = "Staff";
+        private const string Employer = "Employer";
+        private const string HashedAccountId = "HashedAccountId";
+        private const string ESF = "ESF";
+        private const string Tier2User = "Tier2User";
+        private const string ConsoleUser = "ConsoleUser";
+        private const string ESS = "ESS";
+        private const string serviceClaimType = "http://service/service";
+        private static readonly ILogger Logger = LogManager.GetCurrentClassLogger();
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: This should be passed in as a configurable option.

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+        public static IAppBuilder UseSupportConsoleAuthentication(this IAppBuilder app, SupportConsoleAuthenticationOptions options)
+        {
+            app.UseCookieAuthentication(new CookieAuthenticationOptions
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: What is app is null?

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            {
+                AuthenticationType = Employer,
+                ExpireTimeSpan = new TimeSpan(0, 10, 0),
+                SlidingExpiration = true
+            });
+
+            var WsFederationAuthenticationOptions = new WsFederationAuthenticationOptions
+            {
+                AuthenticationType = Staff,
+                Wtrealm = options.AdfsOptions.Wtrealm,
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: What is options is null?

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    {
+        public ADFSOptions AdfsOptions { get; set; }
+    }
+
+    public class ADFSOptions
+    {
+        public string Wtrealm { get; set; }
+        public string MetadataAddress { get; set; }
+        public string Wreply { get; set; }
+        public string BaseUrl { get; set; }
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: Some of these fields are not used and are not ADFSoptions.

File: src/SFA.DAS.EmployerAccounts.Web/Startup.cs [hunk 5]
```diff
@@ -35,10 +35,11 @@ namespace SFA.DAS.EmployerAccounts.Web
     public class Startup
     {
         private static readonly ILogger Logger = LogManager.GetCurrentClassLogger();
-        private const string AccountDataCookieName = "sfa-das-employerapprenticeshipsservice-employeraccount";
+        private const string AccountDataCookieName = "sfa-das-employerapprenticeshipsservice-employeraccount";        
+        private const string TempState = "TempState";
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Startup.cs]: This constant is named after it's value.  It doesn't tell you what it is.

File: src/SFA.DAS.EmployerAccounts.Web/Startup.cs [hunk 6]
```diff
@@ -49,57 +50,50 @@ public void Configuration(IAppBuilder app)
...
+            app.UseSupportConsoleAuthentication(new SupportConsoleAuthenticationOptions 
+            { 
+                AdfsOptions = new ADFSOptions
+                {
+                   
+                    MetadataAddress = config.AdfsMetadata , 
+                    Wreply = config.EmployerAccountsBaseUrl , 
+                    Wtrealm = config.EmployerAccountsBaseUrl ,
+                    BaseUrl = config.Identity.BaseAddress,
+                    ClientId = config.Identity.ClientId,
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Startup.cs]: ClientId, ClientSecret  etc... are not required for the supportConsoleAuthentication

File: src/SFA.DAS.EmployerAccounts.Web/Startup.cs [hunk 7]
```diff
@@ -49,57 +50,50 @@ public void Configuration(IAppBuilder app)
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Startup.cs]: Shouldn't app.SetDefaultSignInAsAuthenticationType(CookieAuthenticationDefaults.AuthenticationType); be removed?

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+    public class ADFSOptions
+    {
+        public string Wtrealm { get; set; }
+        public string MetadataAddress { get; set; }
+        public string Wreply { get; set; }
+        public string BaseUrl { get; set; }      
+        public string Scopes { get; set; }
+        public bool UseCertificate { get; set; }
+        public ILogger Logger { get;  set; }
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: Is logger an ADFS configurable item?

Where are Scopes and UseCertificate used when configuring ADFS?

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                var lastName = claimsIdentity.Claims.SingleOrDefault(x => x.Type == ClaimTypes.Surname)?.Value;
+                var userEmail = claimsIdentity.Claims.Single(x => x.Type == ClaimTypes.Upn).Value;
+
+                claimsIdentity.AddClaim(new Claim(ClaimTypes.Email, userEmail));
+                claimsIdentity.AddClaim(new Claim(ClaimTypes.Name, $"{firstName} {lastName}"));
+                claimsIdentity.AddClaim(new Claim(ClaimTypes.NameIdentifier, userEmail));
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex, "IDAMS Authentication Callback Error");
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: > What happens if an exception is thrown and this method returns. Will the authorisation fail at a later stage?

i have tested with few scenarios  1. Added the claim which is not available in the ClaimsIdentity went to exception and still proceed and worked 2. Removed few claims (basically it  authorise as long as with the required claim)

[the remark this answers is not in the document]
Comment (Stephek) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: Should it authorise if the claim is missing?

[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: Added Exception  if the UPN not available in the claims.

**Comment by VasanthaKasirajan3008** [comment]
Pls review

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+                    claimsIdentity.AddClaim(new Claim(ClaimTypes.Role, ConsoleUser));
+                }
+                else if (notification.AuthenticationTicket.Identity.HasClaim(serviceClaimType, ESS))
+                {
+                    Logger.Debug("Adding ConsoleUser Role");
+                    claimsIdentity.AddClaim(new Claim(ClaimTypes.Role, ConsoleUser));
+                }
+                else
+                {
+                    throw new SecurityTokenValidationException();
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: > Why log , then throw an exception and then log again in the exception handler and throw again? Could just throw the exception here with the message and log the error in the exception handler.

Logging when Service Claim Type not available then throw the SecurityValidationException,
Logging When UPN not available then throw SecurityValidationException
based on your suggestion , i have changed final exception to generic one.

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+                var claimsIdentity = notification.AuthenticationTicket.Identity;
+
+                Logger.Debug("Authentication Properties", new Dictionary<string, object>
+                {
+                    {"claims", JsonConvert.SerializeObject(claimsIdentity.Claims.Select(x =>new {x.Value, x.ValueType, x.Type}))},
+                    {"authentication-type", claimsIdentity.AuthenticationType},
+                    {"role-type", claimsIdentity.RoleClaimType}
+                });
+
+                if (notification.AuthenticationTicket.Identity.HasClaim(serviceClaimType, ESF))
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: done

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+                {
+                    Logger.Error("Upn not available in the claims to identify UserEmail");
+                    throw new SecurityTokenValidationException();
+                }
+                
+            }
+            catch (Exception ex)
+            {
+                Logger.Error(ex, "IDAMS Authentication Callback Error");
+                throw new SecurityTokenValidationException();
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: > Don't throw a new exception you will loose the stack trace for where the original exception was thrown. Just throw the original exception.

Changed to original exception

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+
+                if (notification.AuthenticationTicket.Identity.HasClaim(serviceClaimType, ESF))
+                {
+                    Logger.Debug("Adding Tier2 Role");
+                    claimsIdentity.AddClaim(new Claim(ClaimTypes.Role, Tier2User));
+
+                    Logger.Debug("Adding ConsoleUser Role");
+                    claimsIdentity.AddClaim(new Claim(ClaimTypes.Role, ConsoleUser));
+                }
+                else if (notification.AuthenticationTicket.Identity.HasClaim(serviceClaimType, ESS))
```
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: > What is ESS? Maybe give a more meaningful name.

ESS - ServiceClaimType for Tier1User

File: src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (VasanthaKasirajan3008) [reply src/SFA.DAS.EmployerAccounts.Web/Extensions/SupportUserExtensions.cs]: > What is ESF? Maybe give a more meaningful name.

ESF -> ServiceClaimType for Tier2User

File: src/SFA.DAS.EmployerAccounts.Web.UnitTests/Extensions/TestAppBuilder.cs [hunk 14]
```diff
@@ -1,24 +1,25 @@
 ﻿using Owin;
 using System;
 using System.Collections.Generic;
-using System.Threading.Tasks;
 
 namespace SFA.DAS.EmployerAccounts.Web.UnitTests.Extensions
 {
     public class TestAppBuilder : IAppBuilder
     {
         public IList<Tuple<object, object[]>> BuildMiddleware;
```
Comment (Stephek) [inline src/SFA.DAS.EmployerAccounts.Web.UnitTests/Extensions/TestAppBuilder.cs]: Delete this.

**Pull request merged**